### PR TITLE
Change cpu32 and M68340 to 16 bit data bus instead of 32

### PR DESCRIPTION
**Pull request by Paul-Arnold**
Freescale cpu32 and M68340 were incorrectly defined as having 32 bit data bus.

Changing to 16 bit required a number of drivers to be changed too as they installed 32 bit read/write handlers.

This was raised in issue #10353

### CODE REVIEW

File: src/devices/machine/68340sim.cpp [hunk 1]
```diff
@@ -344,89 +344,117 @@ void m68340_cpu_device::m68340_internal_sim_ports_w(offs_t offset, uint8_t data)
...
+		case m68340_sim::REG_AM_CS3:  data = sim.m_am[3];  break;
+		case m68340_sim::REG_BA_CS3:  data = sim.m_ba[3];  break;
 
 		default:
-			logerror("%08x m68340_internal_sim_r %08x, (%08x)\n", m_ppc, offset*4,mem_mask);
+			logerror("%08x m68340_internal_sim_r %08x, (%08x)\n", m_ppc, offset*2,mem_mask);
 	}
 
-	return 0x00000000;
+	return (BIT(offset,1) ? data : (data >> 16)) & 0xffff;
```
Comment (cuavas) [inline src/devices/machine/68340sim.cpp:371]: Is this correct, or should it be testing the last significant bit?

File: src/devices/machine/68340sim.cpp [hunk 2]
```diff
@@ -344,89 +344,117 @@ void m68340_cpu_device::m68340_internal_sim_ports_w(offs_t offset, uint8_t data)
...
 }
 
-void m68340_cpu_device::m68340_internal_sim_cs_w(offs_t offset, uint32_t data, uint32_t mem_mask)
+void m68340_cpu_device::m68340_internal_sim_cs_w(offs_t offset, uint16_t data, uint16_t mem_mask)
 {
+#define COMBINE_DATA32(varptr)            (*(varptr) = (*(varptr) & ~mem_mask32) | (data32 & mem_mask32))
```
Comment (cuavas) [inline src/devices/machine/68340sim.cpp:376]: Please avoid macros. Use a lambda if you really need to.

File: src/mame/bfm/bfm_swp.cpp [hunk 3]
```diff
@@ -192,8 +187,8 @@ INPUT_PORTS_END
 
 void bfm_swp_state::machine_start()
 {
-	m_cpuregion = (uint32_t*)memregion( "maincpu" )->base();
-	m_mainram = make_unique_clear<uint32_t[]>(0x10000);
+	m_cpuregion = (uint16_t*)memregion( "maincpu" )->base();
```
Comment (cuavas) [inline src/mame/bfm/bfm_swp.cpp:190]: You should really use a `required_region_ptr<uint16_t>` for this.

**Review by cuavas — COMMENTED** [review]
This should probably be left to settle until after the freeze this weekend at least.

Do all CPU32 cores have a 16-bit data bus, or are there other variants? What about the MC68302 MPC? IIRC that embeds a CPU32 core, and some variants support an 8-bit external data bus.

How well have you tested the drivers using this device? IIRC nokia/dbox.cpp uses it, too - does it need adjustments?

**Comment by Paul-Arnold** [comment]
> This should probably be left to settle until after the freeze this weekend at least.
> 
> Do all CPU32 cores have a 16-bit data bus, or are there other variants? What about the MC68302 MPC? IIRC that embeds a CPU32 core, and some variants support an 8-bit external data bus.
> 
> How well have you tested the drivers using this device? IIRC nokia/dbox.cpp uses it, too - does it need adjustments?
The dbox driver was OK it doesn't do any 32 bit read/write mapping.
I ran all drivers using the 68340 and modified each one in turn to fix the problems...all of which is irellivent now...
The 68302 doesn't appear to be CPU32 (according to the datahsheet I have) but searching for that threw up a reference to another device which is cpu32 based with 32 bit data bus :(  That device doesn't show on the NXP web page listing cpu32 based devices, grrrrr!

OK, so this obviously wasn't the way to fix issue. Different method required.

Probably best to delete this PR if that's possible ?

**Comment by cuavas** [comment]
I'm not saying this PR isn't useful, I'm asking questions. If there are CPU32 variants with other bus widths, the base class may need to be more flexible.

**Comment by ajrhacker** [comment]
It's true that some CPU32-based controllers do have the full D0-D31 bus; those versions were usually billed as CPU32+. However, a bigger issue is that CPU32 devices have 32-bit internal configuration registers which should not be reduced to 16 bits.

The CPU32 bus structure is clearly based on that of the MC68020. MC68340 retains both DSACK0 and DSACK1 pins for dynamic bus sizing, so one may even connect 8-bit memory (or, more likely, some sort of programmable peripheral device) to part of the address space. MAME doesn't really handle this well: it requires all RAM and ROM areas, whether internal or external, to have the full bus width, even though this is frequently not the case. However, the MC68340 manual specifically points out that 32-bit memory can in fact be connected externally for single-cycle DMA transfers.

**Comment by Paul-Arnold** [comment]
> However, the MC68340 manual specifically points out that 32-bit memory can in fact be connected externally for single-cycle DMA transfers.

I agree it does say that, but it only exposes d0-d15 to the outside world so I've no idea how that'd work.

> It's true that some CPU32-based controllers do have the full D0-D31 bus; those versions were usually billed as CPU32+. However, a bigger issue is that CPU32 devices have 32-bit internal configuration registers which should not be reduced to 16 bits.

OK, wasn't aware of any in the 68340 but maybe I didn't spot them. I would've expected MBAR to be one of those but it's shown as having word addresses for each half. Even the dma controller shows all it's registers as word addressable which I wouldn't expect.

Not sure where this leaves this change. Is a cpu32 core with 16-bit data just wrong ?

**Comment by smf-** [comment]
As double-address 32-bit dma requires external circuitry and
single-address 32-bit dma doesn't, then I might guess an undocumented
re-use of 16 of the address bits.

**Comment by Paul-Arnold** [comment]
cpu32 has a 16bit data bus both internally and externally.
cpu32+ has 32bit data bus throughout.

Providing two classes, cpu32 and cpu32plus would seem the way to go ?

File: src/mame/bfm/bfm_swp.cpp (hunk 3 above)
Comment (Paul-Arnold) [reply src/mame/bfm/bfm_swp.cpp:190]: OK, will change although I only changed it's size from 32 to 16. It was already there before.

**Comment by Paul-Arnold** [comment]
Just noticed after pushing latest changes that there's a shed load of 68K changes recently submitted which probably affects this ;(

**Comment by galibert** [comment]
I just moved things around, splitting into multiple files, adapting should be trivial, I hope.

**Comment by Paul-Arnold** [comment]
> I just moved things around, splitting into multiple files, adapting should be trivial, I hope.
I like the split files :)
I just don't know what to do now that the PR is showing conflicting files :(

**Comment by galibert** [comment]
Brutal but easy way, go another clone, do the final form of the change in it, copy/pasting stuff around, commit it in one commit, git diff to be sure you're happy with it, then push -f it to your branch.  The PR will update.

**Comment by galibert** [comment]
If the CI passes I think that works for me.

**Pull request merged**